File: src/python/nlp/context_handler.py
"""
Context handler for maintaining conversational state and resolving references
"""

from typing import Dict, List, Optional, Any
from .types import (
    ConversationContext, 
    ContextReference, 
    MentionedEntities, 
    ConversationPreferences,
    PreviousQuery
)

class ContextHandler:
    def __init__(self, max_context_queries: int = 10):
        self.contexts: Dict[str, ConversationContext] = {}
        self.max_context_queries = max_context_queries

    async def get_context(self, session_id: str, user_id: str) -> ConversationContext:
        """Get or create context for a session"""
        context_key = f"{user_id}:{session_id}"
        
        if context_key not in self.contexts:
            self.contexts[context_key] = self._create_new_context(session_id, user_id)

        return self.contexts[context_key]

    async def update_context(
        self, 
        session_id: str, 
        update: Dict[str, Any]
    ) -> None:
        """Update context with new query information"""
        context_key = f"anonymous:{session_id}"  # Simplified for now
        
        if context_key not in self.contexts:
            return  # No context to update

        context = self.contexts[context_key]
        
        # Add to query history
        previous_query = PreviousQuery(
            query=update.get('query', ''),
            intent=update.get('intent'),
            timestamp=update.get('timestamp'),
            result=update.get('result', None)
        )
        context.previous_queries.append(previous_query)

        # Keep only recent queries
        if len(context.previous_queries) > self.max_context_queries:
            context.previous_queries = context.previous_queries[-self.max_context_queries:]

        # Update mentioned entities
        entities = update.get('entities', {})
        if entities.get('table_name'):
            if entities['table_name'] not in context.mentioned_entities.tables:
                context.mentioned_entities.tables.append(entities['table_name'])
            context.current_table = entities['table_name']

        if entities.get('record_id'):
            context.current_record_id = entities['record_id']
            if entities['record_id'] not in context.mentioned_entities.records:
                context.mentioned_entities.records.append(entities['record_id'])

        if entities.get('base_id'):
            context.current_base_id = entities['base_id']

        # Update preferences if detected
        query = update.get('query', '')
        if 'español' in query or 'en español' in query:
            context.preferences.language = 'es'
        elif 'english' in query or 'en inglés' in query:
            context.preferences.language = 'en'

    async def resolve_reference(
        self, 
        session_id: str, 
        reference: str, 
        reference_type: str
    ) -> ContextReference:
        """Resolve contextual references like 'esa tabla' or 'ese registro'"""
        context_key = f"anonymous:{session_id}"
        context = self.contexts.get(context_key)

        if not context:
            return ContextReference(
                reference_type=reference_type,
                reference=reference,
                resolved=False,
                confidence=0.0
            )

        # Resolve specific references
        if reference_type == 'table':
            if reference in ['esa tabla', 'la tabla', 'esta tabla']:
                if context.current_table:
                    return ContextReference(
                        reference_type='table',
                        reference=context.current_table,
                        resolved=True,
                        confidence=0.9,
                        alternatives=context.mentioned_entities.tables
                    )

            # Search in recent queries
            for query_data in reversed(context.previous_queries):
                if context.current_table:
                    return ContextReference(
                        reference_type='table',
                        reference=context.current_table,
                        resolved=True,
                        confidence=0.7,
                        alternatives=context.mentioned_entities.tables
                    )

        elif reference_type == 'record':
            if reference in ['ese registro', 'el registro', 'este registro']:
                if context.current_record_id:
                    return ContextReference(
                        reference_type='record',
                        reference=context.current_record_id,
                        resolved=True,
                        confidence=0.9
                    )

        elif reference_type == 'base':
            if reference in ['esta base', 'la base']:
                if context.current_base_id:
                    return ContextReference(
                        reference_type='base',
                        reference=context.current_base_id,
                        resolved=True,
                        confidence=0.9
                    )

        # If cannot resolve, return original reference with alternatives
        return ContextReference(
            reference_type=reference_type,
            reference=reference,
            resolved=False,
            confidence=0.0,
            alternatives=self._get_alternatives(reference_type, context)
        )

    def _create_new_context(self, session_id: str, user_id: str) -> ConversationContext:
        """Create a new conversation context"""
        return ConversationContext(
            session_id=session_id,
            user_id=user_id,
            current_base_id=None,
            current_table=None,
            current_record_id=None,
            previous_queries=[],
            mentioned_entities=MentionedEntities(),
            preferences=ConversationPreferences()
        )

    def _get_alternatives(self, reference_type: str, context: ConversationContext) -> List[str]:
        """Get alternative references for unresolved references"""
        if reference_type == 'table':
            return context.mentioned_entities.tables
        elif reference_type == 'field':
            return context.mentioned_entities.fields
        elif reference_type == 'record':
            return context.mentioned_entities.records
        elif reference_type == 'base':
            return [context.current_base_id] if context.current_base_id else []
        return []

    async def clear_context(self, session_id: str) -> None:
        """Clear context for a session"""
        context_key = f"anonymous:{session_id}"
        self.contexts.pop(context_key, None)

    async def get_context_history(self, session_id: str) -> List[Dict[str, Any]]:
        """Get query history for a session"""
        context_key = f"anonymous:{session_id}"
        context = self.contexts.get(context_key)
        return [
            {
                'query': pq.query,
                'intent': pq.intent.value,
                'timestamp': pq.timestamp,
                'result': pq.result
            }
            for pq in context.previous_queries
        ] if context else []

    async def get_current_entities(self, session_id: str) -> Dict[str, Any]:
        """Get current entities mentioned in conversation"""
        context_key = f"anonymous:{session_id}"
        context = self.contexts.get(context_key)
        
        if not context:
            return {
                'current_table': None,
                'current_record_id': None,
                'current_base_id': None,
                'mentioned_tables': [],
                'mentioned_fields': [],
                'mentioned_records': []
            }

        return {
            'current_table': context.current_table,
            'current_record_id': context.current_record_id,
            'current_base_id': context.current_base_id,
            'mentioned_tables': context.mentioned_entities.tables,
            'mentioned_fields': context.mentioned_entities.fields,
            'mentioned_records': context.mentioned_entities.records,
            'recent_queries': context.previous_queries[-3:]  # Last 3 queries
        }

    async def set_current_table(self, session_id: str, table_name: str) -> None:
        """Set the current table for context"""
        context_key = f"anonymous:{session_id}"
        context = self.contexts.get(context_key) or self._create_new_context(session_id, 'anonymous')
        
        context.current_table = table_name
        if table_name not in context.mentioned_entities.tables:
            context.mentioned_entities.tables.append(table_name)
        
        self.contexts[context_key] = context

    async def set_current_record(self, session_id: str, record_id: str) -> None:
        """Set the current record for context"""
        context_key = f"anonymous:{session_id}"
        context = self.contexts.get(context_key) or self._create_new_context(session_id, 'anonymous')
        
        context.current_record_id = record_id
        if record_id not in context.mentioned_entities.records:
            context.mentioned_entities.records.append(record_id)
        
        self.contexts[context_key] = context

    async def set_current_base(self, session_id: str, base_id: str) -> None:
        """Set the current base for context"""
        context_key = f"anonymous:{session_id}"
        context = self.contexts.get(context_key) or self._create_new_context(session_id, 'anonymous')
        
        context.current_base_id = base_id
        self.contexts[context_key] = context

File: src/typescript/app/nlp/context-handler.ts
/**
 * Context handler for maintaining conversation state and resolving references
 */

import { 
  ConversationContext, 
  PreviousQuery, 
  MentionedEntities, 
  ConversationPreferences,
  ContextReference
} from '../types/nlp';

export class ContextHandler {
  private contexts: Map<string, ConversationContext> = new Map();
  private maxContextQueries: number;

  constructor(maxContextQueries: number = 10) {
    this.maxContextQueries = maxContextQueries;
  }

  public async getContext(sessionId: string, userId: string): Promise<ConversationContext> {
    let context = this.contexts.get(sessionId);
    
    if (!context) {
      context = this.createNewContext(sessionId, userId);
      this.contexts.set(sessionId, context);
    }
    
    return context;
  }


  public async updateContext(
    sessionId: string, 
    queryData: { 
      query: string; 
      intent: string; 
      timestamp: Date; 
      entities: any; 
      result?: any 
    }
  ): Promise<void> {
    const context = this.contexts.get(sessionId);
    if (!context) {
      throw new Error(`Context not found for session: ${sessionId}`);
    }

    // Add new query to history
    const previousQuery: PreviousQuery = {
      query: queryData.query,
      intent: queryData.intent as any, // Will be properly typed
      timestamp: queryData.timestamp,
      result: queryData.result
    };

    context.previousQueries.push(previousQuery);

    // Limit history size
    if (context.previousQueries.length > this.maxContextQueries) {
      context.previousQueries = context.previousQueries.slice(-this.maxContextQueries);
    }

    // Update mentioned entities if available
    if (queryData.entities) {
      const entities = queryData.entities;
      
      if (entities.tableName && typeof entities.tableName === 'string') {
        const tableName = entities.tableName;
        if (!context.mentionedEntities.tables.includes(tableName)) {
          context.mentionedEntities.tables.push(tableName);
        }
        context.currentTable = tableName;
      }

      if (entities.recordId && typeof entities.recordId === 'string') {
        const recordId = entities.recordId;
        context.currentRecordId = recordId;
        if (!context.mentionedEntities.records.includes(recordId)) {
          context.mentionedEntities.records.push(recordId);
        }
      }

      if (entities.baseId && typeof entities.baseId === 'string') {
        const baseId = entities.baseId;
        context.currentBaseId = baseId;
      }

      if (entities.fieldName && typeof entities.fieldName === 'string') {
        const fieldName = entities.fieldName;
        if (!context.mentionedEntities.fields.includes(fieldName)) {
          context.mentionedEntities.fields.push(fieldName);
        }
      }
    }

    this.contexts.set(sessionId, context);
  }

  public async resolveReference(
    sessionId: string, 
    reference: string, 
    type: string
  ): Promise<ContextReference> {
    const context = this.contexts.get(sessionId);
    if (!context) {
      return {
        referenceType: type,
        reference: '',
        resolved: false,
        confidence: 0
      };
    }

    const lowerReference = reference.toLowerCase();

    if (type === 'table') {
      // Resolve table references
      if (lowerReference.includes('esa tabla') || lowerReference.includes('la tabla')) {
        if (context.currentTable) {
          return {
            referenceType: 'table',
            reference: context.currentTable,
            resolved: true,
            confidence: 0.9
          };
        }
      }

      // Try to match with mentioned tables
      const mentionedTables = context.mentionedEntities.tables;
      if (mentionedTables.length > 0) {
        const lastTable = mentionedTables[mentionedTables.length - 1];
        return {
          referenceType: 'table',
          reference: lastTable,
          resolved: true,
          alternatives: mentionedTables.slice(-3),
          confidence: 0.7
        };
      }
    }

    if (type === 'record') {
      // Resolve record references
      if (lowerReference.includes('ese registro') || lowerReference.includes('el registro')) {
        if (context.currentRecordId) {
          return {
            referenceType: 'record',
            reference: context.currentRecordId,
            resolved: true,
            confidence: 0.9
          };
        }
      }

      // Try to match with mentioned records
      const mentionedRecords = context.mentionedEntities.records;
      if (mentionedRecords.length > 0) {
        const lastRecord = mentionedRecords[mentionedRecords.length - 1];
        return {
          referenceType: 'record',
          reference: lastRecord,
          resolved: true,
          alternatives: mentionedRecords.slice(-3),
          confidence: 0.7
        };
      }
    }

    if (type === 'base') {
      if (context.currentBaseId) {
        return {
          referenceType: 'base',
          reference: context.currentBaseId,
          resolved: true,
          confidence: 0.9
        };
      }
    }

    return {
      referenceType: type,
      reference: '',
      resolved: false,
      confidence: 0
    };
  }

  public async getCurrentEntities(sessionId: string): Promise<any> {
    const context = this.contexts.get(sessionId);
    if (!context) {
      return {
        currentTable: null,
        currentRecordId: null,
        currentBaseId: null,
        mentionedEntities: {
          tables: [],
          fields: [],
          records: []
        },
        recentQueries: []
      };
    }




    return {
      currentTable: context.currentTable ?? null,
      currentRecordId: context.currentRecordId ?? null,
      currentBaseId: context.currentBaseId ?? null,
      mentionedEntities: context.mentionedEntities,
      recentQueries: context.previousQueries.slice(-5).map(q => ({
        query: q.query,
        intent: q.intent,
        timestamp: q.timestamp
      }))
    };
  }

  public async clearContext(sessionId: string): Promise<void> {
    this.contexts.delete(sessionId);
  }

  public async updatePreferences(
    sessionId: string, 
    preferences: Partial<ConversationPreferences>
  ): Promise<void> {
    const context = this.contexts.get(sessionId);
    if (context) {
      context.preferences = { ...context.preferences, ...preferences };
      this.contexts.set(sessionId, context);
    }
  }

  public async getContextualTableSuggestion(sessionId: string): Promise<string | null> {
    const context = this.contexts.get(sessionId);
    if (!context) return null;

    // Suggest current table if available
    if (context.currentTable) {
      return context.currentTable;
    }

    // Suggest most recently mentioned table
    const mentionedTables = context.mentionedEntities.tables;
    if (mentionedTables.length > 0) {
      return mentionedTables[mentionedTables.length - 1];
    }

    return null;
  }

  public async getContextualRecordSuggestion(sessionId: string): Promise<string | null> {
    const context = this.contexts.get(sessionId);
    if (!context) return null;

    // Suggest current record if available
    if (context.currentRecordId) {
      return context.currentRecordId;
    }

    // Suggest most recently mentioned record
    const mentionedRecords = context.mentionedEntities.records;
    if (mentionedRecords.length > 0) {
      return mentionedRecords[mentionedRecords.length - 1];
    }

    return null;
  }

  public async findRecentQueryByIntent(
    sessionId: string, 
    intent: string, 
    limit: number = 3
  ): Promise<PreviousQuery[]> {
    const context = this.contexts.get(sessionId);
    if (!context) return [];

    return context.previousQueries
      .filter(query => query.intent === intent)
      .slice(-limit);
  }

  public async setCurrentContext(
    sessionId: string, 
    table?: string, 
    recordId?: string, 
    baseId?: string
  ): Promise<void> {
    const context = this.contexts.get(sessionId);
    if (context) {
      if (table) context.currentTable = table;
      if (recordId) context.currentRecordId = recordId;
      if (baseId) context.currentBaseId = baseId;
      this.contexts.set(sessionId, context);
    }
  }


  private createNewContext(sessionId: string, userId: string): ConversationContext {
    return {
      sessionId,
      userId,
      currentBaseId: null,
      currentTable: null,
      currentRecordId: null,
      previousQueries: [],
      mentionedEntities: {
        tables: [],
        fields: [],
        records: []
      },
      preferences: {
        language: 'es', // Default to Spanish
        dateFormat: 'YYYY-MM-DD',
        responseFormat: 'natural'
      }
    };
  }

  // Utility method to get context summary
  public getContextSummary(sessionId: string): any {
    const context = this.contexts.get(sessionId);
    if (!context) return null;

    return {
      sessionId: context.sessionId,
      userId: context.userId,
      currentTable: context.currentTable,
      currentRecordId: context.currentRecordId,
      currentBaseId: context.currentBaseId,
      queryCount: context.previousQueries.length,
      mentionedTables: context.mentionedEntities.tables,
      mentionedFields: context.mentionedEntities.fields,
      mentionedRecords: context.mentionedEntities.records,
      preferences: context.preferences
    };
  }
}

